src/pycountry/__init__.py

```diff
@@ -186,12 +204,12 @@ def parent(self):
         return subdivisions.get(code=self.parent_code)
 
 
-class Subdivisions(pycountry.db.Database):
+class Subdivisions(pycountry.db.Database[SubdivisionHierarchy]):
     # Note: subdivisions can be hierarchical to other subdivisions. The
     # parent_code attribute is related to other subdivisions, *not*
     # the country!
 
-    data_class = SubdivisionHierarchy
+    factory = SubdivisionHierarchy
```

Comment:
@terencehonles thanks so much for this PR. I was pulling my hair out trying to figure out how to get the records within a database class to return their respective type. 

The one record type that I'm not sure makes sense is the class for records in Subdivisions. Currently a record like `Alabama` returns a class type of `SubdivisionHierarchy`. Which doesn't follow the rest of the naming conventions where the Database Container Class (i.e. languages) is named for the plural version of the Data Instance Class (i.e. language) it contains. 

What do you think about this implementation to make this more intuitive to end users?

```
class SubdivisionHierarchy(pycountry.db.Data):
    def __init__(self, **kw):
        if "parent" in kw:
            kw["parent_code"] = kw["parent"]
        else:
            kw["parent_code"] = None
        super().__init__(**kw)
        self.country_code = self.code.split("-")[0]
        if self.parent_code is not None:
            # Split the parent_code to check if the country_code is already present
            parts = self.parent_code.split("-")
            if parts[0] != self.country_code:
                self.parent_code = f"{self.country_code}-{self.parent_code}"

    @property
    def country(self):
        return countries.get(alpha_2=self.country_code)

    @property
    def parent(self):
        if not self.parent_code:
            return None
        return subdivisions.get(code=self.parent_code)


class Subdivision(SubdivisionHierarchy):
    pass


class Subdivisions(pycountry.db.Database[Subdivision]):
    # Note: subdivisions can be hierarchical to other subdivisions. The
    # parent_code attribute is related to other subdivisions, *not*
    # the country!

    factory = Subdivision
```

Reply:
I'll have to think about this a little. I realize now that I re-exported `Subdivision` but it's not actually used(?) (and I just flagged this for myself adding the tests that @nschimmoller requested). You're suggesting `Subdivision` be a subclass of `SubdivisionHierarchy`, but should we just get rid of the class? The name isn't super important, and I will need to look at the type definition for `Subdivision.search_fuzy` since I believe it may not be correct (like some of the others).

Reply:
@terencehonles I think you're correct that the `Subdivision` class is not currently used. 

#### Context
--------------
I tried to add the same type hinting you're adding for instances of the containers. I believe prior to this commit the only instance that would properly return their own Class name was `Country`. In an attempt to reverse engineer how `Country` instances were properly storing the intended name I tried to mimic building a `Subdivision` class in `pycountry.db`. As you can see in #175 then Christian, then cleaned up some of the code I was working on to try and add this functionality. 

Since all tests passed, I incorrectly assumed this was implemented correctly. I think again a learning experience as all of us as maintainers are learning to work together. It seems as though incorrectly implementing this is a great example of Cunningham's Law 🤣

So I really do appreciate your willingness to help contribute to `pycountry`! 

#### Naming
--------------
> should we just get rid of the [Subdivision] class?

It does seem that having both `SubdivisionHierarchy` and `Subdivision` may be redundant. I think I would be open to the idea of removing the existing `Subdivision` class and then renaming `SubdivisionHierarchy` to `Subdivision` to achieve the same desired outcome where the instance of the container is the singular version of the plural container. 

> The name isn't super important

While yes, the name isn't super important, I do think a standard naming convention helps end users from a ease of use. For example having `pycountry.countries.search_fuzzy()` and `pycountry.subdivisions.fuzzy_search()` would be confusing for end users. I too so think breaking the naming convention for instances of containers could be confusing. 

> Need to look at the type definition of Subdivision.search_fuzzy()

Yes currently hinted as a `list` containing type `Subdivisions` it actually is a list containing type `SubdivisionHierarchy`. Type hinting for this will depend upon decisions on prior two points